src/PusherCrypto.php

```diff
@@ -0,0 +1,145 @@
+<?php
+
+namespace Pusher;
+
+class PusherCrypto
+{
+    private $encryption_key = '';
+
+    // The prefix any e2e channel must have
+    const ENCRYPTED_PREFIX = 'private-encrypted-';
+
+    // The exact length the user specified key must be
+    const SECRET_KEY_LENGTH = 32;
+
+    /**
+     * Checks if a given channel is an encrypted channel.
+     *
+     * @param string $channel the name of the channel
+     *
+     * @return bool true if channel is an encrypted channel
+     */
+    public static function is_encrypted_channel($channel)
+    {
+        return substr($channel, 0, strlen(self::ENCRYPTED_PREFIX)) === self::ENCRYPTED_PREFIX;
+    }
+
+    /**
+     * Initialises a PusherCrypto instance.
+     *
+     * @param string $encryption_key the SECRET_KEY_LENGTH key that will be used for key derivation.
+     */
+    public function __construct($encryption_key)
+    {
+        if (function_exists('sodium_crypto_secretbox')) {
+            if (strlen($encryption_key) === SODIUM_CRYPTO_SECRETBOX_KEYBYTES) {
+                $this->encryption_key = $encryption_key;
+
+                return;
+            } else {
+                throw new PusherException('Your end to end encryption key must be 32 chars long');
+            }
+        }
+
+        throw new PusherException('To use end to end encryption, you must either be using PHP 7.2 or greater or have installed the libsodium-php extension for php < 7.2.');
+    }
+
+    /**
+     * Decrypts a given event.
+     *
+     * @param object $event an object that has an encrypted data property and a channel property.
+     *
+     * @return object the event with a decrypted payload, or false if decryption was unsuccessful.
+     */
+    public function decrypt_event($event)
+    {
+        $encrypted_payload = explode(':', $event->data);
```

Comment:
Since this data sent from the outside world, I'd feel more confident if we were stricter about validating the event data. i.e. that it contains exactly two `:`s.

Reply:
The only time data will hit this is through the webhook stuff, which has already been validated as coming from Pusher servers. Is this necessary still?

Reply:
Hmm, that is true. I'd still prefer this was validated here.. It could be useful in the rare event that we update the protocol -- the user could get a more helpful error message than a generic decryption failure. The same goes for if the payload became corrupted somehow.